### ISSUE TITLE
mypy doesn't detect use before declaration when on the same line and there is a function call

### ISSUE THREAD

**Issue by mikolajz** [python/mypy#15644]
**Bug Report**

mypy doesn't detect use before declaration when:
* the use is on the same line as the declaration.
* there is a function call around the used-before-declaration variable

**To Reproduce**

[https://mypy-play.net/?mypy=latest&python=3.11&gist=0d661a8817115f77bed9a655fe21fba1]

**Expected Behavior**
A "Name "i" is not defined" error is raised - the code fails at runtime.

**Actual Behavior**
No errors reported

Note:
* replacing `int(...)` with a different function (e.g., a user-defined one) also produces the bug as long as the argument of the function is `int`.
* Removing the function call (`i: int = i`) makes mypy properly produce the error "Name "i" is not defined".

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: Python 3.10.6

**Comment by ikonst** [comment]
Expressions also seem to suppress the check, e.g.
```python
i: int = i * i
j: int = (j,)[0]
```
